group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java

```diff
@@ -262,6 +267,29 @@ public HeartbeatResponseData genericGroupHeartbeat(
         );
     }
 
+    public CoordinatorResult<DeleteGroupsResponseData.DeletableGroupResultCollection, Record> deleteGroups(
+            RequestContext context,
+            List<String> groupIds
+    ) throws ApiException {
+
+        CoordinatorResult<DeleteGroupsResponseData.DeletableGroupResultCollection, Record> groupDeleteCoordinatorResult =
+                groupMetadataManager.groupDelete(context, groupIds);
+
+        List<String> validGroupIds = new ArrayList<>();
+        for (DeleteGroupsResponseData.DeletableGroupResult result : groupDeleteCoordinatorResult.response()) {
+            if (result.errorCode() == Errors.NONE.code()) {
+                validGroupIds.add(result.groupId());
+            }
+        }
+
+        CoordinatorResult<OffsetDeleteResponseData, Record> deleteOffsetCoordinatorResult =
+                offsetMetadataManager.deleteAllOffsets(context, validGroupIds);
+
+        final List<Record> records = groupDeleteCoordinatorResult.records();
+        records.addAll(deleteOffsetCoordinatorResult.records());
+        return new CoordinatorResult<>(records, groupDeleteCoordinatorResult.response());
```

Comment:
I have a few comments regarding this piece of code:
1. I think that we should write the tombstones for the offsets before the ones for the group.
2. It is a bit strange to return a CoordinatorResult from `deleteAllOffsets` and to ignore it. It would be better to pass the list of records to the method and to let the method populate it if the deletion is accepted. I would also remove the response as we don't need it.
3. The `validGroupIds` is a bit weird. How about iterating over the group ids here? Then, you can call the various methods from the manages to validate, delete offsets and finally delete the group. If there is an error, you can directly populate the response with it.

Reply:
I rearranged this part. Now we loop over the group ids and process them one by one -- validate, populate record list with offset tombstones, add group tombstone, and add response.